devito/symbolics/extended_sympy.py

```diff
@@ -383,26 +382,37 @@ class Cast(UnaryOp):
     Symbolic representation of the C notation `(type)expr`.
     """
 
-    _base_typ = ''
+    __rargs__ = ('base', )
+    __rkwargs__ = ('stars', 'dtype')
 
-    __rkwargs__ = ('stars',)
-
-    def __new__(cls, base, stars=None, **kwargs):
-        # Attempt simplifcation
+    def __new__(cls, base, dtype=None, stars=None, **kwargs):
         # E.g., `FLOAT(32) -> 32.0` of type `sympy.Float`
         try:
-            return sympify(eval(cls._base_typ)(base))
+            base = sympify(dtype(base))
         except (NameError, SyntaxError):
-            # E.g., `_base_typ` is "char" or "unsigned long"
+            # E.g., `dtype` is "char" or "unsigned long"
             pass
-        except TypeError:
+        except (ValueError, TypeError):
             # `base` ain't a number
             pass
 
         obj = super().__new__(cls, base)
-        obj._stars = stars
+        obj._stars = stars or ''
+        obj._dtype = cls.__process_dtype__(dtype)
         return obj
 
+    @classmethod
+    def __process_dtype__(cls, dtype):
+        if isinstance(dtype, str):
```

Comment:
I don't think you need this `if` as long as it will be processed fine inside the try-except below

As such, imho you don't need this extra method. You may just move the logic inside `__new__` logically insulating it from the rest with a couple of blank lines